### ISSUE TITLE
UnusedVariableRule bugs

### ISSUE THREAD

**Issue by rudiedirkx** [Slamdunk/phpstan-extensions#68]
I used your `UnusedVariableRule` because apparently I suck at programming :D and it found some good cases, but it also found some invalid cases, and more importantly it did not found the my original bug where my unused variables extension search started.

Important false negative:

```php
$var = $input;
if ($someThing) {
    $var = str_replace('a', 'b', $var);
}
```

`$var` is still unused, because it's not only input of str_replace, but also output. I don't know why it's uncaught, I haven't checked your code in detail, and I'm new at PhpStan/PhpParser so I won't find it anyway.

Less important false positives:

```php
$var = 123;
return get_defined_vars();
```

```php
$var = 123;
return compact('var');
```

```php
$nums = [1, 2, 3];
$var = 123;
$vars = array_map(fn($num) => $num + $var, $nums);
```

In those 3 cases $var is used.

I'm gonna give fixing them a shot, but I expect little. The false positives are annoying, but ignorable. The false negative is way more important, but I feel also much more difficult.

**Comment by rudiedirkx** [comment]
I made a different version, with `NodeTraverser`, and many concrete scenarios :-( https://github.com/rudiedirkx/phpstan-extra/blob/master/src/Php/UnusedVariablesRule.php
